bin/forklift

```diff
@@ -0,0 +1,273 @@
+            'forklift_name': self.name,
+            'forklift_boxes': boxes,
+        }
+        yield play('localhost', variables, ['forklift'], False)
+
+        groups = sorted_groups(self.hosts.keys())
+
+        for segment in self.segments:
+            if segment.common_roles:
+                variables = {
+                    'katello_repositories_version': segment.installer.katello,
+                    'foreman_repositories_version': segment.installer.foreman,
+                    'puppet_repositories_version': segment.installer.puppet,
+                    'katello_repositories_use_koji': True,
+                    # TODO: 3.3 has foreman_repositories_use_koji=true
+                    'foreman_installer_skip_installer': True,
+                }
+                yield play([h.name for h in self.hosts.values()], variables, segment.common_roles)
+
+            for group in groups:
+                variables = segment.hosts[group]['variables']
+                if group == 'proxy':
+                    variables['foreman_proxy_content_server'] = self.hosts['server'].name
+
+                yield play(self.hosts[group].name, variables, segment.hosts[group]['roles'])
+
+        if segment.bats_tests:
+            yield play(self.hosts['server'].name, {'bats_tests': segment.bats_tests}, ['bats'])
+
+    @property
+    def playbook(self):
+        return list(self._plays)
+
+    def write(self, path):
+        with path.open('w') as fp:
+            yaml.safe_dump(self.playbook, fp, allow_unicode=True, default_flow_style=False)
+
+
+def sorted_groups(groups):
+    order = ('server', 'proxy')
+    def get_group_index(group):
+        try:
+            return order.index(group)
+        except ValueError:
+            return len(order)
+    return sorted(groups, key=get_group_index)
+
+
+def get_segments(config):
+    default_pipelines = config['pipelines']
+
+    for version in config['installers']:
+        installer = Installer(version['foreman'], version['katello'], version['puppet'],
+                              version['boxes'])
+
+        pipelines = deepcopy(default_pipelines)
+        if 'pipelines' in version:
+            merge(version['pipelines'], pipelines)
+
+        for product, pipeline in pipelines.items():
+            yield Segment(
+                product=product,
+                version=getattr(installer, product),
+                installer=installer,
+                common_roles=pipeline.get('common_roles', []),
+                bats_tests=pipeline.get('bats_tests', []),
+                hosts=pipeline.get('hosts', {}),
+            )
+
+
+def get_hosts(product, hosts, versions, box):
+    return {group: Host('pipeline-{}-{}-{}-{}'.format(product, group, versions, box), box, options['memory'])
+            for group, options in hosts.items()}
+
+
+def get_pipeline(config, product, versions, box):
+    segments = [s for s in get_segments(config) if s.product == product and s.version in versions]
```

Comment:
Generally not a fan either. I had ```segment``` before but ended up choosing this to limit line length. pep8 doesn't explicitly mention this. I can change ```get_segments``` to do the filtering instead.